src/encodings/levels.rs

```diff
@@ -64,120 +97,352 @@ impl LevelEncoder {
+      },
+      LevelDecoder::BIT_PACKED { bit_width, ref mut num_values, ref mut bit_packed_decoder } => {
+        assert!(num_values.is_some(), "No data set for decoding");
+        // when extracting values from bit reader, it might return more values than left because of
+        // padding to a full byte, we use num_values to track precise number of values.
+        // TODO: Use get_batch() for bit packed decoder
+        let mut i = 0;
+        let len = cmp::min(num_values.unwrap(), buffer.len());
+        while i < len {
+          if let Some(value) = bit_packed_decoder.get_value::<i16>(bit_width as usize) {
+            buffer[i] = value;
+            i += 1;
+          } else {
+            break;
+          }
+        }
+        // update current values
+        *num_values = num_values.map(|x| x - i);
+        Ok(i)
+      },
+    }
   }
 }
 
 #[cfg(test)]
 mod tests {
   use super::*;
+  use util::test_common::random_numbers_range;
 
-  #[test]
-  fn test_roundtrip_one() {
-    let max_level = 1;
-    let levels = vec![0, 1, 1, 1, 1, 0, 0, 0, 0, 1];
-    let max_buffer_size = LevelEncoder::max_buffer_size(Encoding::RLE, max_level, levels.len());
-    let mut encoder = LevelEncoder::new(Encoding::RLE, max_level, vec![0; max_buffer_size]);
+  fn test_internal_roundtrip(encoding: Encoding, levels: &[i16], max_level: i16) {
+    let max_buffer_size = LevelEncoder::max_buffer_size(encoding, max_level, levels.len());
+    let mut encoder = LevelEncoder::new(encoding, max_level, vec![0; max_buffer_size]);
     encoder.put(&levels).expect("put() should be OK");
     let encoded_levels = encoder.consume().expect("consume() should be OK");
 
-    let mut decoder = LevelDecoder::new(Encoding::RLE, max_level);
-    decoder.set_data(ByteBufferPtr::new(encoded_levels));
+    let mut decoder = LevelDecoder::new(encoding, max_level);
+    decoder.set_data(levels.len(), ByteBufferPtr::new(encoded_levels));
     let mut buffer = vec![0; levels.len()];
     let num_decoded = decoder.get(&mut buffer).expect("get() should be OK");
     assert_eq!(num_decoded, levels.len());
     assert_eq!(buffer, levels);
   }
 
+  fn test_internal_roundtrip_incremental(encoding: Encoding, levels: &[i16], max_level: i16) {
+    let max_buffer_size = LevelEncoder::max_buffer_size(encoding, max_level, levels.len());
+    let mut encoder = LevelEncoder::new(encoding, max_level, vec![0; max_buffer_size]);
+    encoder.put(&levels).expect("put() should be OK");
+    let encoded_levels = encoder.consume().expect("consume() should be OK");
+
+    let mut decoder = LevelDecoder::new(encoding, max_level);
+    decoder.set_data(levels.len(), ByteBufferPtr::new(encoded_levels));
+    let mut buffer = vec!();
+    let mut total_decoded = 0;
+    let mut safe_stop = levels.len() * 2; // still terminate in case of issues in the code
+    // perform incremental read until all bytes are read
+    while safe_stop > 0 {
+      safe_stop -= 1;
+      let mut tmp = vec![0; 1];
```

Comment:
not sure if `tmp` is necessary - we can just use `&mut buffer[total_decoded..total_decoded+1]`